### ISSUE TITLE
Is python-fire not compatible with numpy??

### ISSUE THREAD

**Issue by simonnier** [google/python-fire#93]
Here is a simple test

```
import numpy as np
import math
import fire

def test(x):
	a=np.array(x)
	return a

def test2(x):
	a=math.sin(x)
	return a

print('direct call')
print(test([1,2]))

print('fire call')
if __name__ == '__main__':
  fire.Fire()
```

save this as test_fire.py

if we run 

`python test_fire.py test2 1`

it has no problem, it outputs

> [1 2]
> direct call
> [1 2]
> fire call
> 0.8414709848078965

However, if we run 

`python test_fire.py test [1,2]`

it outputs whole lot of things. What is wrong?

> [1 2]
> direct call
> [1 2]
> fire call
> Type:            ndarray
> String form:     [1 2]
> File:            d:\intelpython35\lib\site-packages\numpy\__init__.py
> Class docstring: ndarray(shape, dtype=float, buffer=None, offset=0,
>         strides=None, order=None)
> 
> An array object represents a multidimensional, homogeneous array
> of fixed-size items.  An associated data-type object describes the
> format of each element in the array (its byte-order, how many bytes it
> occupies in memory, whether it is an integer, a floating point number,
> or something else, etc.)
> 
> Arrays should be constructed using `array`, `zeros` or `empty` (refer
> to the See Also section below).  The parameters given here refer to
> a low-level method (`ndarray(...)`) for instantiating an array.
> 
> For more information, refer to the `numpy` module and examine the
> the methods and attributes of an array.
> 
> Parameters
> ----------
> (for the __new__ method; see Notes below)
> 
> shape : tuple of ints
>     Shape of created array.
> dtype : data-type, optional
>     Any object that can be interpreted as a numpy data type.
> buffer : object exposing buffer interface, optional
>     Used to fill the array with data.
> offset : int, optional
>     Offset of array data in buffer.
> strides : tuple of ints, optional
>     Strides of data in memory.
> order : {'C', 'F'}, optional
>     Row-major (C-style) or column-major (Fortran-style) order.
> 
> Attributes
> ----------
> T : ndarray
>     Transpose of the array.
> data : buffer
>     The array's elements, in memory.
> dtype : dtype object
>     Describes the format of the elements in the array.
> flags : dict
>     Dictionary containing information related to memory use, e.g.,
>     'C_CONTIGUOUS', 'OWNDATA', 'WRITEABLE', etc.
> flat : numpy.flatiter object
>     Flattened version of the array as an iterator.  The iterator
>     allows assignments, e.g., ``x.flat = 3`` (See `ndarray.flat` for
>     assignment examples; TODO).
> imag : ndarray
>     Imaginary part of the array.
> real : ndarray
>     Real part of the array.
> size : int
>     Number of elements in the array.
> itemsize : int
>     The memory use of each array element in bytes.
> nbytes : int
>     The total number of bytes required to store the array data,
>     i.e., ``itemsize * size``.
> ndim : int
>     The array's number of dimensions.
> shape : tuple of ints
>     Shape of the array.
> strides : tuple of ints
>     The step-size required to move from one element to the next in
>     memory. For example, a contiguous ``(3, 4)`` array of type
>     ``int16`` in C-order has strides ``(8, 2)``.  This implies that
>     to move from element to element in memory requires jumps of 2 bytes.
>     To move from row-to-row, one needs to jump 8 bytes at a time
>     (``2 * 4``).
> ctypes : ctypes object
>     Class containing properties of the array needed for interaction
>     with ctypes.
> base : ndarray
>     If the array is a view into another array, that array is its `base`
>     (unless that array is also a view).  The `base` array is where the
>     array data is actually stored.
> 
> See Also
> --------
> array : Construct an array.
> zeros : Create an array, each element of which is zero.
> empty : Create an array, but leave its allocated memory unchanged (i.e.,
>         it contains "garbage").
> dtype : Create a data-type.
> 
> Notes
> -----
> There are two modes of creating an array using ``__new__``:
> 
> 1. If `buffer` is None, then only `shape`, `dtype`, and `order`
>    are used.
> 2. If `buffer` is an object exposing the buffer interface, then
>    all keywords are interpreted.
> 
> No ``__init__`` method is needed because the array is fully initialized
> after the ``__new__`` method.
> 
> Examples
> --------
> These examples illustrate the low-level `ndarray` constructor.  Refer
> to the `See Also` section above for easier ways of constructing an
> ndarray.
> 
> First mode, `buffer` is None:
> 
> >>> np.ndarray(shape=(2,2), dtype=float, order='F')
> array([[ -1.13698227e+002,   4.25087011e-303],
>        [  2.88528414e-306,   3.27025015e-309]])         #random
> 
> Second mode:
> 
> >>> np.ndarray((2,), buffer=np.array([1,2,3]),
> ...            offset=np.int_().itemsize,
> ...            dtype=int) # offset = 1*itemsize, i.e. skip first element
> array([2, 3])
> Length:          2
> 
> Usage:           python_fire.py test '[1,2]'
>                  python_fire.py test '[1,2]' T
>                  python_fire.py test '[1,2]' all
>                  python_fire.py test '[1,2]' any
>                  python_fire.py test '[1,2]' argmax
>                  python_fire.py test '[1,2]' argmin
>                  python_fire.py test '[1,2]' argpartition
>                  python_fire.py test '[1,2]' argsort
>                  python_fire.py test '[1,2]' astype
>                  python_fire.py test '[1,2]' base
>                  python_fire.py test '[1,2]' byteswap
>                  python_fire.py test '[1,2]' choose
>                  python_fire.py test '[1,2]' clip
>                  python_fire.py test '[1,2]' compress
>                  python_fire.py test '[1,2]' conj
>                  python_fire.py test '[1,2]' conjugate
>                  python_fire.py test '[1,2]' copy
>                  python_fire.py test '[1,2]' ctypes
>                  python_fire.py test '[1,2]' cumprod
>                  python_fire.py test '[1,2]' cumsum
>                  python_fire.py test '[1,2]' data
>                  python_fire.py test '[1,2]' diagonal
>                  python_fire.py test '[1,2]' dot
>                  python_fire.py test '[1,2]' dtype
>                  python_fire.py test '[1,2]' dump
>                  python_fire.py test '[1,2]' dumps
>                  python_fire.py test '[1,2]' fill
>                  python_fire.py test '[1,2]' flags
>                  python_fire.py test '[1,2]' flat
>                  python_fire.py test '[1,2]' flatten
>                  python_fire.py test '[1,2]' getfield
>                  python_fire.py test '[1,2]' imag
>                  python_fire.py test '[1,2]' item
>                  python_fire.py test '[1,2]' itemset
>                  python_fire.py test '[1,2]' itemsize
>                  python_fire.py test '[1,2]' max
>                  python_fire.py test '[1,2]' mean
>                  python_fire.py test '[1,2]' min
>                  python_fire.py test '[1,2]' nbytes
>                  python_fire.py test '[1,2]' ndim
>                  python_fire.py test '[1,2]' newbyteorder
>                  python_fire.py test '[1,2]' nonzero
>                  python_fire.py test '[1,2]' partition
>                  python_fire.py test '[1,2]' prod
>                  python_fire.py test '[1,2]' ptp
>                  python_fire.py test '[1,2]' put
>                  python_fire.py test '[1,2]' ravel
>                  python_fire.py test '[1,2]' real
>                  python_fire.py test '[1,2]' repeat
>                  python_fire.py test '[1,2]' reshape
>                  python_fire.py test '[1,2]' resize
>                  python_fire.py test '[1,2]' round
>                  python_fire.py test '[1,2]' searchsorted
>                  python_fire.py test '[1,2]' setfield
>                  python_fire.py test '[1,2]' setflags
>                  python_fire.py test '[1,2]' shape
>                  python_fire.py test '[1,2]' size
>                  python_fire.py test '[1,2]' sort
>                  python_fire.py test '[1,2]' squeeze
>                  python_fire.py test '[1,2]' std
>                  python_fire.py test '[1,2]' strides
>                  python_fire.py test '[1,2]' sum
>                  python_fire.py test '[1,2]' swapaxes
>                  python_fire.py test '[1,2]' take
>                  python_fire.py test '[1,2]' tobytes
>                  python_fire.py test '[1,2]' tofile
>                  python_fire.py test '[1,2]' tolist
>                  python_fire.py test '[1,2]' tostring
>                  python_fire.py test '[1,2]' trace
>                  python_fire.py test '[1,2]' transpose
>                  python_fire.py test '[1,2]' var
>                  python_fire.py test '[1,2]' view

**Comment by dbieber** [comment]
Thanks for raising this issue.

**Short answer**
Yes, this is a problem.
You can get around it by running: `python test_fire.py test [1,2] tolist`

**Longer answer**
Since ndarrays are objects and don't fall into any of Fire's special cases, the [help screen for the ndarray is shown](https://github.com/google/python-fire/blob/e76a10c0511ea4485b57543ccb966e7778213279/fire/core.py#L221).

The simplest way to fix this would be to change the way Fire renders ndarrays by adding a [special case here](https://github.com/google/python-fire/blob/e76a10c0511ea4485b57543ccb966e7778213279/fire/core.py#L202).
However, we don't want to add numpy as a dependency of fire, just to be able to check for ndarrays.

So perhaps a better (often requested) option is to allow people to extend / modify the way that Fire serializes it's outputs. We don't have this available yet, but it's a good idea if we can do it well.

**Comment by balabi0github** [comment]
@dbieber Thank you so much for reply. numpy is so widely used, hope this issue will be solved in the near future.

**Comment by discdiver** [comment]
Is this still an issue?

**Comment by dbieber** [comment]
This is resolved in v0.2.0, which will be release later today July 26.
